File: internal/skirk/protocol.go
package skirk

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
)

const (
	envelopeMagic = "SKB1"
	envelopeVer   = byte(1)
	sessionIDLen  = 16
	keyLen        = 32
	headerLen     = 39
	maxSequence   = uint64(1<<56 - 1)
	DirectionUp   = byte(1)
	DirectionDown = byte(2)
	FlagData      = byte(0)
	FlagFinal     = byte(1)
)

type Envelope struct {
	SessionID    [16]byte
	Direction    byte
	Sequence     uint64
	Flags        byte
	PlaintextLen uint32
	Ciphertext   []byte
}

func RandomSecret() (string, error) {
	key := make([]byte, keyLen)
	if _, err := rand.Read(key); err != nil {
		return "", err
	}
	return "base64:" + base64.StdEncoding.EncodeToString(key), nil
}

func NewSessionID() ([16]byte, error) {
	var sid [16]byte
	_, err := rand.Read(sid[:])
	return sid, err
}

func ParseSessionID(value string) ([16]byte, error) {
	if value == "" {
		return NewSessionID()
	}
	raw, err := hex.DecodeString(value)
	if err != nil {
		return [16]byte{}, err
	}
	if len(raw) != sessionIDLen {
		return [16]byte{}, fmt.Errorf("session id must be %d bytes / 32 hex chars", sessionIDLen)
	}
	var sid [16]byte
	copy(sid[:], raw)
	return sid, nil
}

func SessionString(sid [16]byte) string {
	return hex.EncodeToString(sid[:])
}

func DeriveKey(secret string) ([]byte, error) {
	value := strings.TrimSpace(secret)
	switch {
	case strings.HasPrefix(value, "hex:"):
		raw, err := hex.DecodeString(value[4:])
		if err != nil {
			return nil, err
		}
		if len(raw) != keyLen {
			return nil, fmt.Errorf("hex key must be %d bytes", keyLen)
		}
		return raw, nil
	case strings.HasPrefix(value, "base64:"):
		raw, err := base64.StdEncoding.DecodeString(value[7:])
		if err != nil {
			return nil, err
		}
		if len(raw) != keyLen {
			return nil, fmt.Errorf("base64 key must be %d bytes", keyLen)
		}
		return raw, nil
	default:
		return hkdfSHA256([]byte(value), []byte("skirk-v1-static-salt"), []byte("skirk-blobq-aead-key"), keyLen), nil
	}
}

func hkdfSHA256(ikm, salt, info []byte, length int) []byte {
	extract := hmac.New(sha256.New, salt)
	extract.Write(ikm)
	prk := extract.Sum(nil)

	var okm []byte
	var previous []byte
	counter := byte(1)
	for len(okm) < length {
		expand := hmac.New(sha256.New, prk)
		expand.Write(previous)
		expand.Write(info)
		expand.Write([]byte{counter})
		previous = expand.Sum(nil)
		okm = append(okm, previous...)
		counter++
	}
	return okm[:length]
}

func Seal(key []byte, sid [16]byte, direction byte, sequence uint64, plaintext []byte, final bool) ([]byte, error) {
	if len(key) != keyLen {
		return nil, fmt.Errorf("key must be %d bytes", keyLen)
	}
	if sequence > maxSequence {
		return nil, errors.New("sequence out of supported nonce range")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	flags := FlagData
	if final {
		flags = FlagFinal
	}
	header := make([]byte, headerLen)
	copy(header[0:4], []byte(envelopeMagic))
	header[4] = envelopeVer
	copy(header[5:21], sid[:])
	header[21] = direction
	header[22] = flags
	binary.BigEndian.PutUint64(header[23:31], sequence)
	binary.BigEndian.PutUint32(header[31:35], uint32(len(plaintext)))
	binary.BigEndian.PutUint32(header[35:39], uint32(len(plaintext)+gcm.Overhead()))
	ciphertext := gcm.Seal(nil, nonce(sid, direction, sequence), plaintext, header)
	return append(header, ciphertext...), nil
}

func OpenEnvelope(key, data []byte) (Envelope, []byte, error) {
	if len(data) < headerLen {
		return Envelope{}, nil, errors.New("envelope too short")
	}
	header := data[:headerLen]
	if !bytes.Equal(header[0:4], []byte(envelopeMagic)) {
		return Envelope{}, nil, errors.New("bad envelope magic")
	}
	if header[4] != envelopeVer {
		return Envelope{}, nil, fmt.Errorf("unsupported envelope version %d", header[4])
	}
	var sid [16]byte
	copy(sid[:], header[5:21])
	direction := header[21]
	flags := header[22]
	sequence := binary.BigEndian.Uint64(header[23:31])
	plaintextLen := binary.BigEndian.Uint32(header[31:35])
	ciphertextLen := binary.BigEndian.Uint32(header[35:39])
	if int(ciphertextLen) != len(data)-headerLen {
		return Envelope{}, nil, errors.New("ciphertext length mismatch")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return Envelope{}, nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return Envelope{}, nil, err
	}
	ciphertext := data[headerLen:]
	plaintext, err := gcm.Open(nil, nonce(sid, direction, sequence), ciphertext, header)
	if err != nil {
		return Envelope{}, nil, err
	}
	if len(plaintext) != int(plaintextLen) {
		return Envelope{}, nil, errors.New("plaintext length mismatch")
	}
	return Envelope{
		SessionID:    sid,
		Direction:    direction,
		Sequence:     sequence,
		Flags:        flags,
		PlaintextLen: plaintextLen,
		Ciphertext:   ciphertext,
	}, plaintext, nil
}

func nonce(sid [16]byte, direction byte, sequence uint64) []byte {
	out := make([]byte, 12)
	copy(out[0:4], sid[:4])
	out[4] = direction
	for i := 0; i < 7; i++ {
		out[11-i] = byte(sequence >> (8 * i))
	}
	return out
}

File: skirk/protocol.py
"""Encrypted chunk envelope used by Skirk mailbox-style transports."""

from __future__ import annotations

import base64
import os
import struct
from dataclasses import dataclass
from enum import IntEnum

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


MAGIC = b"SKB1"
VERSION = 1
HEADER = struct.Struct("!4sB16sBBQII")
SESSION_LEN = 16
KEY_LEN = 32
MAX_SEQ = (1 << 56) - 1


class Direction(IntEnum):
    UP = 1
    DOWN = 2


class Flags(IntEnum):
    DATA = 0
    FINAL = 1


@dataclass(frozen=True)
class Envelope:
    session_id: bytes
    direction: Direction
    sequence: int
    flags: int
    plaintext_len: int
    ciphertext: bytes

    def object_name(self) -> str:
        return f"{self.session_id.hex()}/{self.direction.name.lower()}/{self.sequence:016x}.skb"


def new_session_id() -> bytes:
    return os.urandom(SESSION_LEN)


def encode_session_id(session_id: bytes) -> str:
    if len(session_id) != SESSION_LEN:
        raise ValueError("session_id must be 16 bytes")
    return session_id.hex()


def decode_session_id(value: str | None) -> bytes:
    if not value:
        return new_session_id()
    raw = bytes.fromhex(value)
    if len(raw) != SESSION_LEN:
        raise ValueError("session id must be 16 bytes / 32 hex chars")
    return raw


def derive_key(secret: str | bytes) -> bytes:
    """Derive a 256-bit AEAD key from a user secret or raw key string."""

    if isinstance(secret, str):
        value = secret.strip()
        if value.startswith("hex:"):
            raw = bytes.fromhex(value[4:])
            if len(raw) != KEY_LEN:
                raise ValueError("hex key must be 32 bytes")
            return raw
        if value.startswith("base64:"):
            raw = base64.b64decode(value[7:])
            if len(raw) != KEY_LEN:
                raise ValueError("base64 key must be 32 bytes")
            return raw
        ikm = value.encode("utf-8")
    else:
        ikm = secret
    return HKDF(
        algorithm=hashes.SHA256(),
        length=KEY_LEN,
        salt=b"skirk-v1-static-salt",
        info=b"skirk-blobq-aead-key",
    ).derive(ikm)


def _nonce(session_id: bytes, direction: Direction, sequence: int) -> bytes:
    if len(session_id) != SESSION_LEN:
        raise ValueError("session_id must be 16 bytes")
    if not 0 <= sequence <= MAX_SEQ:
        raise ValueError("sequence out of supported nonce range")
    return session_id[:4] + bytes([int(direction)]) + sequence.to_bytes(7, "big")


def seal(
    *,
    key: bytes,
    session_id: bytes,
    direction: Direction,
    sequence: int,
    plaintext: bytes,
    final: bool = False,
) -> bytes:
    flags = int(Flags.FINAL if final else Flags.DATA)
    header = HEADER.pack(
        MAGIC,
        VERSION,
        session_id,
        int(direction),
        flags,
        sequence,
        len(plaintext),
        # AES-GCM appends a 16-byte tag to the plaintext-length ciphertext.
        len(plaintext) + 16,
    )
    ciphertext = AESGCM(key).encrypt(_nonce(session_id, direction, sequence), plaintext, header)
    return header + ciphertext


def open_envelope(*, key: bytes, data: bytes) -> tuple[Envelope, bytes]:
    if len(data) < HEADER.size:
        raise ValueError("envelope too short")
    magic, version, session_id, direction_raw, flags, sequence, plaintext_len, ciphertext_len = HEADER.unpack(
        data[: HEADER.size]
    )
    if magic != MAGIC:
        raise ValueError("bad envelope magic")
    if version != VERSION:
        raise ValueError(f"unsupported envelope version {version}")
    if ciphertext_len != len(data) - HEADER.size:
        raise ValueError("ciphertext length mismatch")
    direction = Direction(direction_raw)
    header = data[: HEADER.size]
    ciphertext = data[HEADER.size :]
    plaintext = AESGCM(key).decrypt(_nonce(session_id, direction, sequence), ciphertext, header)
    if len(plaintext) != plaintext_len:
        raise ValueError("plaintext length mismatch")
    return (
        Envelope(
            session_id=session_id,
            direction=direction,
            sequence=sequence,
            flags=flags,
            plaintext_len=plaintext_len,
            ciphertext=ciphertext,
        ),
        plaintext,
    )


def random_secret() -> str:
    return "base64:" + base64.b64encode(os.urandom(KEY_LEN)).decode("ascii")
